### PR TITLE
Only respect ALL environments when not in parallel

### LINKED ISSUE #2167: Parallel mode with -e ALL fails
Using ``-p`` together with ``-e ALL`` runs **all environments in all subprocesses**. Using ``-e <list,of,all,envs>`` on the other hand works as expected (each subprocess runs one environment).

Minimal example:

``tox.ini``

```ini
[tox]
envlist = py

[testenv]
deps = requests
skip_install = true
commands = python -V

[testenv:lint]
commands = python -c "print('Hello lint')"

[testenv:dev]
commands = python -c "print('Hello dev')"
```

Environment:

```bash
$ tox --version
3.24.2 imported from ~\.local\pipx\venvs\tox\lib\site-packages\tox\__init__.py
$ ~/.local/pipx/venvs/tox/Scripts/python -V
Python 3.9.6
```

Correct command:

```bash
$ tox -p -o -e py,lint,dev
py create: ~\Dev\tmp\test_tox\.tox\py
dev create: ~\Dev\tmp\test_tox\.tox\dev
lint create: ~\Dev\tmp\test_tox\.tox\lint
py installdeps: requests
dev installdeps: requests
lint installdeps: requests
py installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
py run-test-pre: PYTHONHASHSEED='937'
py run-test: commands[0] | python -V
dev installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
dev run-test-pre: PYTHONHASHSEED='633'
dev run-test: commands[0] | python -c 'print('"'"'Hello dev'"'"')'
lint installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
lint run-test-pre: PYTHONHASHSEED='783'
lint run-test: commands[0] | python -c 'print('"'"'Hello lint'"'"')'
Python 3.9.7
✔ OK py in 8.858 seconds
Hello dev
Hello lint
✔ OK dev in 8.929 seconds
✔ OK lint in 8.936 seconds
____________________________________________________________ summary ____________________________________________________________
  py: commands succeeded
  lint: commands succeeded
  dev: commands succeeded
  congratulations :)
```

Incorrect behaviour:

```bash
$ tox -p -o -e ALL
py installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
py run-test-pre: PYTHONHASHSEED='672'
py run-test: commands[0] | python -V
py installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
py run-test-pre: PYTHONHASHSEED='524'
py run-test: commands[0] | python -V
py installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
py run-test-pre: PYTHONHASHSEED='977'
py run-test: commands[0] | python -V
Python 3.9.7
Python 3.9.7
Python 3.9.7
lint installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
lint installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
lint run-test-pre: PYTHONHASHSEED='524'
lint run-test: commands[0] | python -c 'print('"'"'Hello lint'"'"')'
lint run-test-pre: PYTHONHASHSEED='672'
lint run-test: commands[0] | python -c 'print('"'"'Hello lint'"'"')'
lint installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
lint run-test-pre: PYTHONHASHSEED='977'
lint run-test: commands[0] | python -c 'print('"'"'Hello lint'"'"')'
Hello lint
Hello lint
Hello lint
dev installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
dev run-test-pre: PYTHONHASHSEED='524'
dev run-test: commands[0] | python -c 'print('"'"'Hello dev'"'"')'
dev installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
dev run-test-pre: PYTHONHASHSEED='672'
dev run-test: commands[0] | python -c 'print('"'"'Hello dev'"'"')'
dev installed: certifi==2021.5.30,charset-normalizer==2.0.4,idna==3.2,requests==2.26.0,urllib3==1.26.6
dev run-test-pre: PYTHONHASHSEED='977'
dev run-test: commands[0] | python -c 'print('"'"'Hello dev'"'"')'
Hello dev
Hello dev
Hello dev
✔ OK dev in 5.777 seconds
✔ OK py in 5.802 seconds
✔ OK lint in 5.812 seconds
____________________________________________________________ summary ____________________________________________________________
  py: commands succeeded
  lint: commands succeeded
  dev: commands succeeded
  congratulations :)
```

Please note, that in the incorrect case, all three echos happen three times.

The issues comes from https://github.com/tox-dev/tox/blob/156d4961dd50ee629c3c4c78c09ba6b1c0da76d3/src/tox/config/__init__.py#L1501, which causes https://github.com/tox-dev/tox/blob/156d4961dd50ee629c3c4c78c09ba6b1c0da76d3/src/tox/config/__init__.py#L1507 to not correctly restrict the ``env_list`` to the environment given by ``os.environ.get(PARALLEL_ENV_VAR_KEY_PRIVATE)``.

### PR DESCRIPTION
**Pull request by guahki**
fixes #2167

## Contribution checklist:

(also see [CONTRIBUTING.rst](../tree/master/CONTRIBUTING.rst) for details)

- [ ] wrote descriptive pull request text
- [x] added/updated test(s)
- [ ] updated/extended the documentation
- [x] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [x] added news fragment in [changelog folder](../tree/master/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`, `breaking`, `doc`, `misc`
  * if PR has no issue: consider creating one first or change it to the PR number after creating the PR
  * "sign" fragment with ```-- by :user:`<your username>`.```
  * please, use full sentences with correct case and punctuation, for example:
    ```rst
    Fixed an issue with non-ascii contents in doctest text files -- by :user:`superuser`.
    ```
  * also see [examples](../tree/master/docs/changelog)
- [ ] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

**Comment by guahki** [comment]
As

1. this is my first contribution to this project and
1. I'm not sure about the correct way to fix this

this is kind of a draft/request PR. I will happily create appropriate changelog when you tell me, this is a sensible way of fixing the problem.

Thanks for the awesome ``tox`` tool!

File: src/tox/config/__init__.py [hunk 1]
```diff
@@ -1498,9 +1498,10 @@ def _getenvdata(self, reader, config):
 
         env_list = []
         envlist_explicit = False
-        if (from_option and "ALL" in from_option) or (
-            not from_option and from_environ and "ALL" in from_environ.split(",")
-        ):
+        if (
+            (from_option and "ALL" in from_option)
+            or (not from_option and from_environ and "ALL" in from_environ.split(","))
+        ) and not os.environ.get(PARALLEL_ENV_VAR_KEY_PRIVATE):
```
Comment (guahki) [inline src/tox/config/__init__.py:1504]: Maybe one just wants to check for

```suggestion
        ) and PARALLEL_ENV_VAR_KEY_PRIVATE not in os.environ:
```

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Can you please add changelog and test for this?

**Comment by guahki** [comment]
I added the changelog.

Unfortunately I do not have the time to write a test right now. I would be happy, if anybody can assist with this.

**Comment by guahki** [comment]
I had a look into it and it wasn't that hard indeed.
Thanks for your patience!

**Review by gaborbernat — APPROVED** [review]

**Pull request merged**